### PR TITLE
update sequence generator to support generation from a language model

### PR DESCRIPTION
**Pull request by amrsharaf**
This is a simple update for the `sequence_generator.py` script to support generation from a language model trained on the target vocabulary. Changes include:

1) Clone `max_positions` for source / target if it's set by a monolingual  LM;
2) Set the encoder output dictionary to None when using LM;
3) Check for None values of encoder output dictionary when calling `reorder_encoder_out`
4) Assert targets are given for adaptive softmax only at training time;
5) Copy attention values only if provided by the decoder model;
6) Squeeze the `probs` tensor instead of `decoder_out` as the adaptive softmax probability function expects 3D inputs.

This is a follow up on the issues: #213 and #212. I'm basically replicating the LM results from the hierarchical story generation paper (https://arxiv.org/abs/1805.04833), but I can't find the LM generation code in the repo. I'd appreciate if @huihuifan can take a look at this, in case it's a replicated or redundant pull request.

### CODE REVIEW

**Comment by huihuifan** [comment]
The experiments from the story paper were actually run with luatorch, but I wanted a PyTorch code release because most people use PyTorch now. @alexeib implemented the convlm model in PyTorch as well (I had it in Lua as well when we originally worked on that paper). Will take a look right now. @alexeib as well?

File: fairseq/data/language_pair_dataset.py [hunk 1]
```diff
@@ -139,6 +139,10 @@ def valid_size(self, index, max_positions):
     def _get_max_positions(self, max_positions):
         if max_positions is None:
             return self.max_source_positions, self.max_target_positions
+        # If max_position is not a sequence (possibly from a monolingual model)
+        # then use the same upper bound for source & target
+        if not hasattr(max_positions, '__len__'):
```
Comment (alexeib) [inline fairseq/data/language_pair_dataset.py]: can we check if its not type(tuple) instead ?

File: fairseq/models/fconv.py [hunk 2]
```diff
@@ -494,8 +494,13 @@ def get_normalized_probs(self, net_output, log_probs, sample):
...
         if self.adaptive_softmax is not None:
-            assert sample is not None and 'target' in sample
-            out = self.adaptive_softmax.get_log_prob(net_output[0], sample['target'])
+            if sample is not None:
+                # Assert enforced only at training time
+                assert 'target' in sample
+                out = self.adaptive_softmax.get_log_prob(net_output[0], sample['target'])
+            else:
+                # No target was provided, setting to None
+                out = self.adaptive_softmax.get_log_prob(net_output[0], target=None)
```
Comment (alexeib) [inline fairseq/models/fconv.py]: can we move "target" out into a variable and then we can just call get_log_prob once unconditionally? (to make refactors easier in the future)

File: fairseq/sequence_generator.py [hunk 3]
```diff
@@ -268,7 +272,7 @@ def get_hypo():
                 for i, model in enumerate(self.models):
                     if isinstance(model.decoder, FairseqIncrementalDecoder):
                         model.decoder.reorder_incremental_state(incremental_states[model], reorder_state)
-                    encoder_outs[i] = model.encoder.reorder_encoder_out(encoder_outs[i], reorder_state)
+                    encoder_outs[i] = model.decoder.reorder_encoder_out(encoder_outs[i], reorder_state)
```
[the remark this answers is not in the document]
Comment (alexeib) [reply fairseq/sequence_generator.py]: i thought they are all on encoder now, which maes sense.

in case of a language model encoder_outs should be empty anyway, maybe we can just check that? and this particular change would probably cause all the encoder-decoder models to crash (e.g. machine translation)

File: fairseq/sequence_generator.py [hunk 4]
```diff
@@ -521,9 +526,10 @@ def _decode_one(self, tokens, model, encoder_out, incremental_states, log_probs)
                 decoder_out = list(model.decoder(tokens, encoder_out, incremental_states[model]))
             else:
                 decoder_out = list(model.decoder(tokens, encoder_out))
-            decoder_out[0] = decoder_out[0][:, -1, :]
             attn = decoder_out[1]
             if attn is not None:
                 attn = attn[:, -1, :]
         probs = model.get_normalized_probs(decoder_out, log_probs=log_probs)
+        probs = probs[:, -1, :]
```
Comment (alexeib) [inline fairseq/sequence_generator.py]: why move it here? its probably much faster to get probs just for the last state than for everything, and then take the last one?

File: fairseq/sequence_generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (huihuifan) [reply fairseq/sequence_generator.py]: oops yes I forgot we agreed to move them all into the encoder

[the remark this answers is not in the document]
Comment (amrsharaf) [reply fairseq/sequence_generator.py]: Yes, this should be `encoder` instead of `decoder`, I'll fix it.

@huihuifan , @alexeib Thank you so much for looking into this. I'll address all the comments and run test to make sure the changes doesn't break generation for translation.

File: fairseq/data/language_pair_dataset.py (hunk 1 above)
Comment (alexeib) [reply fairseq/data/language_pair_dataset.py]: sorry i meant check if type is int here

Comment (amrsharaf) [reply fairseq/data/language_pair_dataset.py]: max_positions was actually a float, I think this check is more general as it capture both tuple and list...

File: fairseq/sequence_generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (huihuifan) [reply fairseq/sequence_generator.py]: thanks!

**Comment by myleott** [comment]
Cool! @alexeib already covered all the comments I'd have. Please also make sure the integration tests pass: `python setup.py test`.
